### PR TITLE
fix `deepEquals` with array holes and accessors

### LINKED ISSUE #2417: Bun test fails `toEqual` test for non-empty array passed through `produce` from `immer`
### What version of Bun is running?

0.5.7

### What platform is your computer?

Linux 5.15.90.1-microsoft-standard-WSL2 x86_64 x86_64

### What steps can reproduce the bug?

Reproducible repo [here](https://github.com/jakeboone02/bun-test-immer-toequal/blob/main/src/testrun.ts#L6-L10). Run `bun test bun.test & bunx jest jest` to see the result of both bun:test and Jest for the same test suite.

Short version:

```ts
import { produce } from "immer";

test('fails in bun but not jest', () => {
  const array1 = [""];
  const array2 = [""];
  expect(produce(array1, a => a)).toEqual(array2);
});
```

### What is the expected behavior?

bun:test's `toEqual` should behave the same as Jest's `toEqual` even when `immer` is involved.

### What do you see instead?

bun:test always fails a `toEqual` comparison between two non-empty arrays if exactly one of the arrays is passed through `immer`'s `produce` function, even if the `produce` function simply returns the draft unchanged.

If the arrays are empty, or if they are _both_ passed through `produce` first, the test will pass in both bun:test and Jest.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by dylan-conway**
fixes #2417

### CODE REVIEW

File: src/bun.js/bindings/bindings.cpp [hunk 1]
```diff
@@ -167,6 +167,22 @@ static bool canPerformFastPropertyEnumerationForIterationBun(Structure* s)
...
 
+JSValue getIndexWithoutAccessors(JSGlobalObject* globalObject, JSObject* obj, uint64_t i)
+{
+    if (obj->canGetIndexQuickly(i)) {
+        return obj->tryGetIndexQuickly(i);
+    }
+
+    PropertySlot slot(obj, PropertySlot::InternalMethodType::Get);
+    if (obj->methodTable()->getOwnPropertySlotByIndex(obj, globalObject, i, slot)) {
+        if (!slot.isAccessor()) {
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/bindings.cpp:178]: are you sure we shouldn't call accessors here?

```js
❯ cat /tmp/a.js
       │ File: /tmp/a.js
   1   │ const obj = {};
   2   │ Object.defineProperty(obj, "0", {get() { return 42; }});
   3   │ console.log(obj[0])

❯ bun /tmp/a.js
42
```

perfectly valid, terrible code

Comment (dylan-conway) [reply src/bun.js/bindings/bindings.cpp:178]: yes, this passes in jest:
```js
const obj1 = {};
Object.defineProperty(obj1, "0", {
  get() {
    return 42;
  },
});
const obj2 = {};
Object.defineProperty(obj2, "0", {
  get() {
    return 44;
  },
});
expect(obj1).toStrictEqual(obj2);
```

Comment (dylan-conway) [reply src/bun.js/bindings/bindings.cpp:178]: ah, but maybe `deepEquals` outside of jest shouldn't behave this way

Comment (Jarred-Sumner) [reply src/bun.js/bindings/bindings.cpp:178]: ah, then lets wait for someone to file an issue before we add more complexity

**Comment by jakeboone02** [comment]
Thanks for addressing this. Hope this comment is useful:  https://github.com/oven-sh/bun/issues/2417#issuecomment-1496725527

**Review by Jarred-Sumner — APPROVED** [review]
LGTM

**Comment by jakeboone02** [comment]
@dylan-conway/@Jarred-Sumner I don't see any tests in here specifically for the `Object.preventExtensions` issue I brought up in https://github.com/oven-sh/bun/issues/2417#issuecomment-1496725527. Are we sure this definitely fixes the issue? Great if it does, I just don't see how the tests cover it.

**Comment by dylan-conway** [comment]
@jakeboone02 Oops, I should've included the repro you had. I tested locally and it is working `Object.preventExtensions` now

**Comment by jakeboone02** [comment]
Excellent. Thanks again!

**Pull request merged**